tpm2_pytss/tsskey.py

```diff
@@ -0,0 +1,241 @@
+from asn1crypto.core import ObjectIdentifier, Sequence, Boolean, OctetString, Integer
+from asn1crypto import pem
+
+
+parent_rsa_template = TPMT_PUBLIC(
+    type=TPM2_ALG.RSA,
+    nameAlg=TPM2_ALG.SHA256,
+    objectAttributes=TPMA_OBJECT.USERWITHAUTH
+    | TPMA_OBJECT.RESTRICTED
+    | TPMA_OBJECT.DECRYPT
+    | TPMA_OBJECT.NODA
+    | TPMA_OBJECT.FIXEDTPM
+    | TPMA_OBJECT.FIXEDPARENT
+    | TPMA_OBJECT.SENSITIVEDATAORIGIN,
+    authPolicy=b"",
+    parameters=TPMU_PUBLIC_PARMS(
+        rsaDetail=TPMS_RSA_PARMS(
+            symmetric=TPMT_SYM_DEF_OBJECT(
+                algorithm=TPM2_ALG.AES,
+                keyBits=TPMU_SYM_KEY_BITS(aes=128),
+                mode=TPMU_SYM_MODE(aes=TPM2_ALG.CFB),
+            ),
+            scheme=TPMT_RSA_SCHEME(scheme=TPM2_ALG.NULL),
+            keyBits=2048,
+            exponent=0,
+        ),
+    ),
+)  # FIXME, use template from tpm2-openssl
+parent_ecc_template = TPMT_PUBLIC(
+    type=TPM2_ALG.ECC,
+    nameAlg=TPM2_ALG.SHA256,
+    objectAttributes=TPMA_OBJECT.USERWITHAUTH
+    | TPMA_OBJECT.RESTRICTED
+    | TPMA_OBJECT.DECRYPT
+    | TPMA_OBJECT.NODA
+    | TPMA_OBJECT.FIXEDTPM
+    | TPMA_OBJECT.FIXEDPARENT
+    | TPMA_OBJECT.SENSITIVEDATAORIGIN,
+    authPolicy=b"",
+    parameters=TPMU_PUBLIC_PARMS(
+        eccDetail=TPMS_ECC_PARMS(
+            symmetric=TPMT_SYM_DEF_OBJECT(
+                algorithm=TPM2_ALG.AES,
+                keyBits=TPMU_SYM_KEY_BITS(aes=128),
+                mode=TPMU_SYM_MODE(aes=TPM2_ALG.CFB),
+            ),
+            scheme=TPMT_ECC_SCHEME(scheme=TPM2_ALG.NULL),
+            curveID=TPM2_ECC.NIST_P256,
+            kdf=TPMT_KDF_SCHEME(scheme=TPM2_ALG.NULL),
+        ),
+    ),
+)  # FIXME, use template from tpm2-openssl
+
+rsa_template = TPMT_PUBLIC(
+    type=TPM2_ALG.RSA,
+    nameAlg=TPM2_ALG.SHA256,
+    objectAttributes=TPMA_OBJECT.USERWITHAUTH
+    | TPMA_OBJECT.SIGN_ENCRYPT
+    | TPMA_OBJECT.DECRYPT
+    | TPMA_OBJECT.NODA
+    | TPMA_OBJECT.FIXEDTPM
+    | TPMA_OBJECT.FIXEDPARENT
+    | TPMA_OBJECT.SENSITIVEDATAORIGIN,
+    authPolicy=b"",
+    parameters=TPMU_PUBLIC_PARMS(
+        rsaDetail=TPMS_RSA_PARMS(
+            symmetric=TPMT_SYM_DEF_OBJECT(algorithm=TPM2_ALG.NULL),
+            scheme=TPMT_RSA_SCHEME(scheme=TPM2_ALG.NULL),
+        ),
+    ),
+)  # FIXME
+ecc_template = TPMT_PUBLIC(
+    type=TPM2_ALG.ECC,
+    nameAlg=TPM2_ALG.SHA256,
+    objectAttributes=TPMA_OBJECT.USERWITHAUTH
+    | TPMA_OBJECT.SIGN_ENCRYPT
+    | TPMA_OBJECT.NODA
+    | TPMA_OBJECT.FIXEDTPM
+    | TPMA_OBJECT.FIXEDPARENT
+    | TPMA_OBJECT.SENSITIVEDATAORIGIN,
+    authPolicy=b"",
+    parameters=TPMU_PUBLIC_PARMS(
+        eccDetail=TPMS_ECC_PARMS(
+            symmetric=TPMT_SYM_DEF_OBJECT(algorithm=TPM2_ALG.NULL),
+            scheme=TPMT_ECC_SCHEME(scheme=TPM2_ALG.NULL),
+            kdf=TPMT_KDF_SCHEME(scheme=TPM2_ALG.NULL),
+        ),
+    ),
+)  # FIXME
+
+loadablekey_oid = ObjectIdentifier("2.23.133.10.1.3")
+
+
+class BooleanOne(Boolean):
```

Comment:
Interesting that the base class ASN1 Boolean didn't have this behavior. What exactly changes between the two?

Reply:
yeah, my understanding of the DER encoding is that as long as it's not 0 it is true, and asn1crypto encodes True as 0xFF while openssl/tpm2-tss-engine encodes True as 0x01.
While it isn't necessary I think it's good if the same input will lead to the same output as tpm2-tss-engine / tpm2-openssl